Fix #4904: Update questions backend to align with Skills construct

aks681 opened the pull request:
The various validations done on a question before its creation has been updated to make sure it aligns with the properties of a question in a skill. 
The rest of the changes to the question construct is to be done after the SkillQuestionLink model has been created.

aks681 committed:
fix #4904

aks681 committed:
fixed lint errors

nithusha21 commented on core/domain/exp_domain.py:
```diff
@@ -1269,7 +1274,8 @@ def __init__(self, content, param_changes, interaction,
         self.classifier_model_id = classifier_model_id
 
-    def validate(self, exp_param_specs_dict, allow_null_interaction):
+    def validate(self, exp_param_specs_dict, allow_null_interaction,
```
Nit: Break after the parenthesis '('. This should fix the lint failure.

aks681 replied:
Done.

DubeySandeep commented on scripts/backend_tests.py:
```diff
@@ -218,18 +218,18 @@ def _convert_to_test_target(path):
         if any([s in root for s in ['.git', 'third_party', 'core/tests']]):
             continue
-        if root.endswith('_test.py'):
+        if root.endswith('exp_domain_test.py'):
```
@aks681, is this expected?

aks681 replied:
Had changed that for testing, removed it.

DubeySandeep commented on scripts/backend_tests.py:
```diff
@@ -218,18 +218,18 @@ def _convert_to_test_target(path):
             if _LOAD_TESTS_DIR in subroot and include_load_tests:
                 for f in files:
-                    if f.endswith('_test.py'):
+                    if f.endswith('exp_domain_test.py'):
```
Ditto.

aks681 replied:
Done

DubeySandeep commented on scripts/backend_tests.py:
```diff
@@ -218,18 +218,18 @@ def _convert_to_test_target(path):
 
             for f in files:
-                if (f.endswith('_test.py') and
+                if (f.endswith('exp_domain_test.py') and
```
Ditto.

aks681 replied:
Done

DubeySandeep commented on core/domain/question_domain.py:
```diff
@@ -147,8 +147,41 @@ def validate(self):
                 'Expected question_data to be a dict, received %s' %
                 self.question_data)
+
+        #return self.question_data['interaction']
```
Do we need this?

aks681 replied:
question_data is a State object, so it has to be a dict , that's why I put the check, just as a precaution. Should it be removed?

BenHenning replied:
I think @DubeySandeep was specifically asking whether we need this comment. Since it doesn't seem to be providing any additional context, can we just remove it?

seanlip replied:
I think @DubeySandeep was asking about the commented-out line 151.

aks681 replied:
Oh, ok, will remove that.

aks681 replied:
Done

aks681 committed:
made review changes

aks681 commented:
@seanlip PTAL

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -1179,10 +1183,11 @@ def validate(self, exp_param_specs_dict):
-            self.default_outcome.validate()
+        if not is_question:
+            for answer_group in self.answer_groups:
+                answer_group.validate(interaction, exp_param_specs_dict)
```
What about validating other aspects of ``AnswerGroup``? Is there anything in the outcome that's worth validating in both cases? Should the ``is_question`` check be delegated to ``Outcome`` to decide how to validate?

aks681 replied:
All of the checks related to answer groups and outcomes exclusive to Question, I am doing in question.validate() itself.

BenHenning commented on scripts/start.sh:
```diff
@@ -64,13 +64,17 @@ fi
     CLEAR_DATASTORE_ARG=""
   fi
+  if [ "$arg" == "--enable_console" ]; then
+    ENABLE_CONSOLE_ARG="--enable_console=true"
```
What is this for & how does it relate to the skills work? Can this be done as a separate PR if it's unrelated?

aks681 replied:
This is not related to this PR in particular. For enabling developer console at localhost:8000 (which was what I used to test Questions as the frontend is not present), this extra argument was required, so I made this change, so that in the future anyone can enable developer console by just adding this argument when running start.sh.

seanlip replied:
Ah. Can you put it in a separate PR then? That way, you can include the rationale for the change in the PR description and it'll be easier to understand when looking back on it in future.

aks681 replied:
Ok, will do that.

aks681 replied:
Done, removed it, will create a separate PR for this alone.

BenHenning commented on core/domain/question_domain.py:
```diff
@@ -147,8 +147,41 @@ def validate(self):
+                'Expected all answer groups to have destination as None.'
+            )
+
+        if (len(interaction['hints']) == 0) | (interaction['solution'] is None):
```
Prefer ``or`` instead of ``|``.

aks681 replied:
Done

seanlip reviewed (changes requested):
Hi @aks681, I feel iffy about this approach. I think passing around an is_question argument everywhere isn't a good idea. It becomes hard to maintain. Also, "is_question" is an intrinsic property of the object and should be modelled as such.

What do you think about, instead, having a Question class which subclasses State?

BenHenning reviewed (changes requested):
Apologies for the late review.

I'm actually not keen on the boolean being passed down in this way because, from the perspective of ``exp_domain``, understanding the nuances of validating a state object (and, especially, an outcome) requires searching for callsites and elsewhere in the code. That may not be clear to other developers, and could be a bit brittle as we change validation invariants over time. My follow-up question is: what aspects of ``Outcome`` are currently validated for exploration states that should not be for questions? Questions having additional validation rules definitely makes sense, but do we need to leak the concept of questions down into the state structure?

If we do need to inhibit some of the validation currently being done in State in the context of questions, there maybe a slightly cleaner way to do it (maybe by introducing a separate validation path for those behaviors). Happy to make further suggestions once I better understand why we can't rely on the existing validate() call in the questions context.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -147,8 +147,41 @@ def validate(self):
             raise utils.ValidationError(
                 'Expected question_data to be a dict, received %s' %
                 self.question_data)
+
```
I can't comment higher up, but maybe we should drop self.title, per the questions design proposal?

Also we might need a new QuestionData object that subclasses State (see other comment). Need to implement some mechanism for migration as well.

seanlip replied:
Actually, scrap that -- @BenHenning's comment got me thinking about whether we actually need a separate object here in the first place. Perhaps we can pull the question/exploration-specific validation checks  higher up, so that they're done in exploration.validate() or question.validate() instead? Then outcome.validate() or state.validate() would only do the checks that are common to both cases.

BenHenning replied:
That sounds good to me. I was only thinking of this from the questions & shared perspective before, but pulling up exploration-specific checks above the state level also seems reasonable and provides parity with questions.

aks681 replied:
With question title dropped, QuestionSummary has only question_id. So, should that also be removed, or should some values of question_data be added to summary?

Ok, I will do the changes to the validations.

aks681 replied:
Then, for creating a default question_data state, shall I create another function in class State (create_default_question_data()), or for this, should I create the new QuestionData object, like the initial comment, and have this function (create_default_question_data()) inside that (in this case, should the interaction and outcome, of question_data, still be objects of their respective classes in exp_domain, or just individual dicts?).

seanlip replied:
Re QuestionSummary: I'm not sure, actually. I haven't fully thought about what QuestionSummary should look like. It would probably contain a summary of the question text but we'll need to figure out how to generate that. For now, I suggest storing the content of the question's state; what do you think?

Re creating a default state, I think a new function in class State is fine. You might even go so far as to rename question_data to 'state' altogether, since it is a state.

aks681 replied:
Ok, I'll create the function in class State.

If the entire question_data is stored, then there is not much difference between class Question and QuestionSummary (only question_id, schema_version and language_code). I thought of just storing interaction name or some field of question_data only, or is a QuestionSummary required, as question object itself doesn't have a lot of information for itself?

aks681 replied:
Done.
Also, I changed the QuestionSummary to just hold question id and content, but when a question is created, there is no function to create its summary as of now (even in develop). Should I add that now or later?

seanlip commented:
Oh, to be clear, for the summary, I meant just the non-interactive content
field of the question. Not the whole question data / state object.

aks681 commented:
Oh, Ok, will do that.

aks681 committed:
Made review changes

aks681 committed:
fixed errors

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into update-question

aks681 committed:
fixed errors

aks681 committed:
removed __enable_console arg

aks681 commented:
@seanlip PTAL. Removed the is_question flag altogether, and moved up some validation checks in class InteractionInstance which was exclusive to an exploration to exploration.validate(). Also, changed QuestionSummary object.

aks681 committed:
fixed lint errors

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1584,11 +1569,45 @@ def create_default_state(
+        interaction = InteractionInstance.create_default_interaction(
+            default_dest_state_name).to_dict()
+
+        interaction = InteractionInstance.from_dict(interaction)
```
Not sure I understand why we do a to_dict() then a from_dict() here? (Isn't the net result of those a no-op, meaning that those operations can be removed?)

aks681 replied:
Oh, I had done that, when the question's default state was also created here, will change that back.

aks681 replied:
Done.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1584,11 +1569,45 @@ def create_default_state(
+            default_dest_state_name).to_dict()
+        solution = {
+            'answer_is_exclusive': False,
+            'correct_answer': 'helloworld!',
```
helloworld! and hello_world is a string should not be defaults here. You don't want them to show up for every new question.

Also the default correct_answer should have the correct type based on the interaction (so to make this clear, maybe have an Interaction.create_default_correct_answer() and add a comment there saying that the type of the default answer must match the interaction -- or better still add a backend test to ensure that those two methods are always and forever consistent). That way if the default interaction ever changes, we guard against errors due to the 'correct answer' still being associated with the old default interaction type.

aks681 replied:
Removed these from exp_domain and added function to create valid question data in the test files, as it is only there, where these are called.
So, there I just gave a sample TextInput interaction.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1584,11 +1569,45 @@ def create_default_state(
+                'audio_translations': {}
+            }
+        }
+        interaction['id'] = 'Continue'
```
This doesn't seem like an interaction that would be used for Questions. It also does not match the type of the default correct answer.

Is there actually a strong reason here to use defaults that are different from the usual State?

aks681 replied:
The main reason was that in the validation for this a correct answer, hints and solutions should be present with no destination state. But come to think of it, when a question is created initially, when a creator goes to the editor window for the first time, it doesn't make sense to have an interaction with hints and solution specified already. Shall I then, remove these validations and have a check just before commit in question_services for hints, solutions and others?

seanlip replied:
This is interesting, actually. I think it depends on when a question is "created" in the backend.

For explorations, the exploration is created in the backend before the exploration editor is even shown. So the "new" version of the exploration must pass validation, since that is "version 1" of the exploration.

For questions, I think the situation is different. When the creator goes to the question editor page and makes a new question, that question should not be stored in the backend yet. It should just be initialized locally as a frontend dict. Only once the creator tries to save it, should the question be sent to the backend and validated.

This has a couple of implications: the result of "create default question" in the backend can just be passed to the frontend without validation as the "starter template" for the user, and validation can be stricter than it would be otherwise.

What do you think of this approach?

aks681 replied:
Yeah, that's fine. In that case, a create_default_question_data is not needed, right? The existing create_default_question() can be used (which returns empty dict as question_data), and the validations can be done just before actually saving the question (before model.commit). Is that ok?

seanlip replied:
Well, depends. Wouldn't your frontend need a populated question dict?

aks681 replied:
Ya, for that, the existing create_default_state would be fine, as validations are not done anyways.

aks681 replied:
Done. For this, I modified the create_default_question, as mentioned (by calling create_default_state for question_data), and added a separate function in only the test files to create a valid question data, as the validation checks are only done before saving the question in the datastore, the default question framework that a creator sees in the beginning need not be a valid question object.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1958,9 +1977,29 @@ def validate(self, strict=False):
+            # InteractionInstance as these tests are exclusive to an interaction
+            # in a state in an exploration, and not for that in a question,
+            # (whereas the validate() in InteractionInstance is also used by
+            # the question_data field).
```
.. used by the State object that constitutes the question_data field.

aks681 replied:
Done.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1958,9 +1977,29 @@ def validate(self, strict=False):
+                except KeyError:
+                    raise utils.ValidationError(
+                        'Invalid interaction id: %s' % state.interaction.id)
+                for answer_group in state.interaction.answer_groups:
```
I think only this line and the next one need to be moved out? Anything that is needed for both cases should go in the common place (under Interaction).

In fact if there's generic validation for answer groups it should still go in the common place. Just pull out the really specific validation that needs exp_param_specs higher up.

It also seems reasonable to add a comment there saying "Validation of XXX is done in either the Exploration or Question class since both of these treat parameters differently."

aks681 replied:
You mean, to have the outcome validation also, inside Interaction? Actually, only the first validation inside outcome (the one where it checks for destination states), is the only one exclusive for explorations, shall I just pull that out, then?

seanlip replied:
Yep.

aks681 replied:
Done, only the first two checks in validate() in class Outcome is now in the validate() function in class Exploration.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -140,16 +135,45 @@ def validate(self):
                 self.question_data)
+
+        at_least_one_correct_answer = False
+        is_dest_specified = False
```
dest_is_specified (to indicate boolean rather than function)

aks681 replied:
Done.

seanlip commented on core/domain/question_domain.py:
```diff
@@ -140,16 +135,45 @@ def validate(self):
+                'Expected at least one answer group to have a correct answer.'
+            )
+
+        if interaction['default_outcome']['dest'] is not None:
```
Put this above the previous block, together with lines 152-153. This grouping makes more sense, I think -- you set up everything, then validate everything.

aks681 replied:
Done.

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -17,7 +17,7 @@
 import utils
 
-from google.appengine.ext import ndb
+from google.appengine.ext import ndb  # pylint: disable=relative-import
```
Did you check this with @apb7? We should avoid adding pylint pragmas where possible, and fix the underlying issue.

apb7 replied:
Hi @seanlip! @aks681 had discussed this with me. For relative-import errors, we disable the check for the corresponding line similar to [these lines](https://github.com/oppia/oppia/blob/develop/constants_test.py#L19:L20) in `./constants_test.py`.  
But I tested it locally just now and I do not face any error or problem without the disable. Also, the develop branch does not have the disable statement and all checks pass. Let me talk to @aks681 once again regarding this. Thanks!

aks681 replied:
Yeah, in travis they pass, but when running pre_submit checks in my local machine, they fail. Even, when I push develop also, this happens.

apb7 replied:
@aks681: All tests pass locally on my machine. Similarly for other developers. Also, this check has not been recently added, it was already enabled in Pylint. So I think it might be a local machine problem, not sure though. Thanks!

aks681 replied:
Ok, removed it.

seanlip replied:
@aks681 -- for reference, how did you resolve this?

seanlip commented:
Took another pass, PTAL.

> Also, I changed the QuestionSummary to just hold question id and content, but when a question is created, there is no function to create its summary as of now (even in develop). Should I add that now or later?

I'm happy with that being done here or in a separate PR. I think since there are some other review changes that need to be made, it's probably more expedient to do them in this PR, but your call.

aks681 committed:
initial changes

aks681 committed:
made review changes

aks681 committed:
minor fix

aks681 committed:
fixed backend test

aks681 commented:
@seanlip PTAL. I added the _create_valid_question_data() function to only the test files as that is the only place where they are actually used. Also, just shifted the first two validations in class Outcome to exploration.validate().

seanlip commented on core/controllers/question_test.py:
```diff
@@ -32,6 +32,41 @@ class QuestionsHandlersTest(test_utils.GenericTestBase):
+            dict. The default question_data dict.
+        """
+        state = exp_domain.State.create_default_state(
+            default_dest_state_name).to_dict()
```
Better to create a valid domain object and then do a single to_dict() at the end. That gives you more validation/guarantees.

(Generally, try to work with domain objects, not dicts.)

aks681 replied:
Done.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1144,9 +1136,15 @@ def get_all_outcomes(self):
         """Validates various properties of the InteractionInstance.
 
+        Args:
+            exp_param_specs_dict: dict. A dict of specified parameters used in
```
shouldn't this be "dict or None", to account for the Question case?

Also maybe in the Question case it makes sense to pass an empty dict? Not sure.

aks681 replied:
I changed the validate call in question.validate() to empty dict.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1978,28 +1951,32 @@ def validate(self, strict=False):
-            # the question_data field).
-            if state.interaction.id is not None:
-                if not isinstance(state.interaction.id, basestring):
+            # Outcome (which are called from the validate() function in class
```
Don't talk about "shifting" in the docstring; new devs won't know what that refers to since they only see the new version.

Something like "The checks below perform validation on the Outcome domain object that is specific to answer groups in explorations, but not questions. This logic is here because the validation checks in the Outcome domain object are used by both explorations and questions." should suffice.

aks681 replied:
Done.

seanlip commented on core/domain/question_domain_test.py:
```diff
@@ -19,12 +19,48 @@
 class QuestionDomainTest(test_utils.GenericTestBase):
     """Tests for Question domain object."""
 
+    def _create_valid_question_data(self, default_dest_state_name):
```
Why duplicate? If you need a setup method common to both tests, then put it in core/tests/test_utils.py. In general try not to duplicate code.

aks681 replied:
Done, moved the function to test_utils.

seanlip commented on core/domain/question_services_test.py:
```diff
@@ -28,6 +28,41 @@
 class QuestionServicesUnitTest(test_utils.GenericTestBase):
     """Test the question services module."""
 
+    def _create_valid_question_data(self, default_dest_state_name):
```
ditto, don't duplicate

aks681 replied:
Done.

seanlip reviewed (commented):
Sorry @aks681, I think I missed this review request. Done another pass!

aks681 committed:
made review changes

aks681 commented:
@seanlip PTAL. Made the review changes. Also, about the import error, I am still getting that, but for now, after making sure run_presubmit_checks pass (except that error), I am pushing with --no-verify. I will look into that.

seanlip commented:
OK, thanks! We are actually looking into removing that flag altogether, so definitely don't depend on it :)

I'll take a look at the PR shortly.

seanlip reviewed (approved):
LGTM. Thanks!
